cassandra/io/eventletreactor.py

```diff
@@ -0,0 +1,213 @@
+
+        self._socket = green.socket.socket(green.socket.AF_INET,
+                                           green.socket.SOCK_STREAM)
+        self._socket.settimeout(1.0)
+        self._socket.connect((self.host, self.port))
+
+        if self.sockopts:
+            for args in self.sockopts:
+                self._socket.setsockopt(*args)
+
+        self._read_watcher = eventlet.spawn(lambda: self.handle_read())
+        self._write_watcher = eventlet.spawn(lambda: self.handle_write())
+
+        self._send_options_message()
+
+    def close(self):
+        with self.lock:
+            if self.is_closed:
+                return
+            self.is_closed = True
+
+        log.debug("Closing connection (%s) to %s" % (id(self), self.host))
+
+        cur_gthread = eventlet.getcurrent()
+
+        if self._read_watcher and self._read_watcher != cur_gthread:
+            self._read_watcher.kill()
+            # Avoid refcycles
+            self._read_watcher = None
+        if self._write_watcher and self._write_watcher != cur_gthread:
+            self._write_watcher.kill()
+            # Avoid refcycles
+            self._write_watcher = None
+        if self._socket:
+            self._socket.close()
+            self._socket = None
+        log.debug("Closed socket to %s" % (self.host,))
+
+        if not self.is_defunct:
+            self.error_all_callbacks(
+                connection.ConnectionShutdown("Connection to %s was closed"
+                                              % self.host))
+            # don't leave in-progress operations hanging
+            self.connected_event.set()
+
+    def handle_write(self):
+        while True:
+            try:
+                next_msg = self._write_queue.get()
+                self._socket.send(next_msg)
+            except green.socket.error as err:
+                log.debug(
+                    "Exception during socket sendall for %s: %s", self, err)
+                self.defunct(err)
+                return  # Leave the write loop
+
+    def handle_read(self):
+        run_select = functools.partial(green.select.select,
+                                       (self._socket,), (), ())
+        while True:
+            try:
+                run_select()
+            except Exception as exc:
+                if not self.is_closed:
+                    log.debug(
+                        "Exception during read select() for %s: %s", self, exc)
+                    self.defunct(exc)
+                return
+
+            try:
+                buf = self._socket.recv(self.in_buffer_size)
+                self._iobuf.write(buf)
+            except green.socket.error as err:
+                if not is_timeout(err):
+                    log.debug(
+                        "Exception during socket recv for %s: %s", self, err)
+                    self.defunct(err)
+                    return  # leave the read loop
+
+            if self._iobuf.tell():
```

Comment:
Is there a reason this logic is here, in handle_read instead of left to connection.process_io_buffer?